pkg/platform/model/vcs.go

```diff
@@ -643,6 +600,38 @@ func CommitInitial(hostPlatform string, langName, langVersion string) (strfmt.UU
 	return res.Payload.CommitID, nil
 }
 
+func isWildcardVersion(version string) bool {
+	return strings.Index(version, "x") >= 0 || strings.Index(version, "X") >= 0
+}
+
+func VersionStringToConstraints(version string) ([]*mono_models.Constraint, error) {
+	if !isWildcardVersion(version) {
+		return []*mono_models.Constraint{{Comparator: "eq", Version: version}}, nil // exact version
+	}
+
+	// Construct version constraints to be >= given version, and < given version's last part + 1.
+	// For example, given a version number of 3.10.x, constraints should be >= 3.10, < 3.11.
+	// Given 2.x, constraints should be >= 2, < 3.
+	constraints := []*mono_models.Constraint{}
+	parts := strings.Split(version, ".")
+	for i, part := range parts {
+		if part != "x" && part != "X" {
+			continue
+		}
+		if i == 0 {
+			return nil, locale.NewInputError("err_version_wildcard_start", "A version number cannot start with a wildcard")
+		}
+		constraints = append(constraints, &mono_models.Constraint{Comparator: "gte", Version: strings.Join(parts[:i], ".")})
+		previousPart, err := strconv.Atoi(parts[i-1])
+		if err != nil {
+			return nil, locale.WrapInputError(err, "err_version_number_expected", "Version parts are expected to be numeric")
+		}
+		parts[i-1] = strconv.Itoa(previousPart + 1)
+		constraints = append(constraints, &mono_models.Constraint{Comparator: "lt", Version: strings.Join(parts[:i], ".")})
+	}
```

Comment:
This does not feel like it should be using mono model structs since we're not sending this to the mono API anymore. Probably makes more sense in the buildplan model?

Reply:
We are using the mono API for `state init` (vcs.go's `CommitInitial()` function). In order to avoid duplication, I've decided to have the string -> constraints list translation happen here in vcs.go. BuildPlanner makes use of this, but translates from mono API constraints to build planner constraints.

Let me know if you'd like me to reverse this. The `state init` work came first, so it has precedence in my mind.